packages/@sanity/schema/src/sanity/validation/types/array.ts

```diff
@@ -106,6 +133,79 @@ export default (typeDef, visitorContext) => {
     )
   }
 
+  const [primitiveTypes, objectTypes] = partition(
```

Comment:
This code is kinda involved with a lot of nested ifs. Is it covered by unit-tests? Would make it a lot less scary to modify in the future.

Reply:
I'm afraid that goes for almost everything in this package. I think it's better addressed by doing a separate quality  improvement iteration on this package (includes adding strict mode, upping test coverage etc).